File: leetcode 668.py
class Solution:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        def check(x):
            bound = x // n
            cnt = bound * n
            for i in range(bound + 1, m + 1):
                cnt += x // i
            return cnt >= k
        l = 0; r = m*n
        while l < r:
            mid = (l + r) >> 1
            if check(mid):
                r = mid
            else:
                l = mid + 1
        return l

File: leetcode 668.cpp
class Solution {
public:
  int findKthNumber(int m, int n, int k) {
    int left = 1, right = m * n;
    auto check = [&](int x) -> bool {
      int bound = x / n, cnt = bound * n;
      for (int i = bound + 1; i <= m; i++)
        cnt += x / i;
      return cnt >= k;
    };
    while (left < right) {
      int mid = (left + right) >> 1;
      if (check(mid))
        right = mid;
      else
        left = mid + 1;
    }
    return left;
  }
};